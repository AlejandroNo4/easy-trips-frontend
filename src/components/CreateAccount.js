import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate } from 'react-router-dom';
import fetchingPost from '../api/fetchingPost';
import FormSignUp from './FormSignUp';
import fetchingGet from '../api/fetchingGet';

const CreateAccount = () => {
  const dispatch = useDispatch();
  const userState = useSelector((state) => state.UIReducer);
  const navigate = useNavigate();

  useEffect(() => {
    const url = 'logged_in';
    const type = 'UI';
    fetchingGet({
      dispatch,
      url,
      type,
    });
    if (userState.user.logged_in === true) navigate('/');
  }, []);

  const initialStateForm = {
    username: '',
    email: '',
    password: '',
    passwordConfirmation: '',
    image: [],
  };
  const [form, updateInput] = useState(initialStateForm);

  const handleChange = (e) => {
    if (e.target.files) updateInput({ ...form, [e.target.name]: e.target.files[0] });
    else updateInput({ ...form, [e.target.name]: e.target.value });
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    const {
      username, email, password, passwordConfirmation, image,
    } = form;

    const formData = {
      user: {
        username,
        email: email.toLowerCase(),
        password,
        password_confirmation: passwordConfirmation,
        image,
      },
    };
    const url = '/users';
    const type = 'UI';
    fetchingPost({
      dispatch,
      url,
      formData,
      navigate,
      type,
    });
  };

  if (userState.loading === true) {
    return <h1>------LOADING...------</h1>;
  }
  return (
    <div>
      <FormSignUp handleChange={handleChange} handleSubmit={handleSubmit} />
    </div>
  );
};

export default CreateAccount;
