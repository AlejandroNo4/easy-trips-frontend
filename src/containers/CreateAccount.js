import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate, Link } from 'react-router-dom';
import fetchingPost from '../api/fetchingPost';
import FormSignUp from '../components/FormSignUp';
import fetchingGet from '../api/fetchingGet';
import { cleanupErrors } from '../actions';

const CreateAccount = () => {
  const dispatch = useDispatch();
  const userState = useSelector((state) => state.UIReducer);
  const navigate = useNavigate();

  useEffect(() => {
    const url = 'logged_in';
    const type = 'UI';
    fetchingGet({
      dispatch,
      url,
      type,
    });
    if (userState.user.logged_in === true) navigate('/');
    return () => {
      dispatch(cleanupErrors());
    };
  }, []);

  const initialStateForm = {
    username: '',
    email: '',
    password: '',
    passwordConfirmation: '',
    image: {},
  };
  const [form, updateInput] = useState(initialStateForm);

  const handleChange = (e) => {
    if (e.target.files) updateInput({ ...form, [e.target.name]: e.target.files[0] });
    else updateInput({ ...form, [e.target.name]: e.target.value });
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    const {
      username, email, password, passwordConfirmation, image,
    } = form;

    const formData = {
      user: {
        username,
        email: email.toLowerCase(),
        password,
        password_confirmation: passwordConfirmation,
        image,
      },
    };
    const url = '/users';
    const type = 'UI';
    fetchingPost({
      dispatch,
      url,
      formData,
      navigate,
      type,
    });
  };

  if (userState.loading === true) {
    return (
      <div className="d-flex flex-column justify-center align-center w-100">
        <h1 className="session-title">Loading...</h1>
      </div>
    );
  }

  let errors;
  if (userState.loading === false) {
    const keys = Object.keys(userState.errors);
    errors = keys.map((key) => `${key}: ${userState.errors[key]}`).join('\n');
  }

  return (
    <div className="bg-no-session d-flex justify-center flex-column align-center no-session-container" data-testid="sign-up-container">
      <p className="error-msg" data-testid="sign-up-errs">{errors}</p>
      <h1 className="session-title">Sign up</h1>
      <p className="session-description text-center">
        Please, create an account.
      </p>
      <FormSignUp
        handleChange={handleChange}
        handleSubmit={handleSubmit}
        imgSelected={form.image}
      />
      <Link to="/" className="link-back">
        Go back
      </Link>
    </div>
  );
};

export default CreateAccount;
