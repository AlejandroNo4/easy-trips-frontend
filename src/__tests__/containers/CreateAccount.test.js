import React from 'react';
import { Provider } from 'react-redux';
import {
  render, screen, cleanup, fireEvent, act,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import axiosMock from 'axios';
import { BrowserRouter } from 'react-router-dom';
import store from '../../store';
import CreateAccount from '../../containers/CreateAccount';

afterEach(cleanup);

const renderWithRedux = (component) => ({
  ...render(
    <Provider store={store}>
      <BrowserRouter>{component}</BrowserRouter>
    </Provider>,
  ),
});

describe('CreateAccount Component', () => {
  beforeEach(() => {
    renderWithRedux(<CreateAccount />);
  });

  it('renders with Redux', async () => {
    const signUpTitle = await screen.findByTestId('sign-up-container');
    expect(signUpTitle).toBeVisible();
  });

  it('matches the snapshot', async () => {
    const signUpTitle = await screen.findByTestId('sign-up-container');
    expect(signUpTitle).toMatchSnapshot();
  });

  it('Renders username and tranfers the value while typing', async () => {
    const username = await screen.findByPlaceholderText('Username');
    expect(username).toBeVisible();
    userEvent.type(username, 'Mario');
    expect(username).toHaveProperty('required', true);
    expect(username).not.toHaveProperty('required', false);
    expect(username).toHaveValue('Mario');
    expect(username).not.toHaveValue('');
  });

  it('Renders email and tranfers the value while typing', async () => {
    const email = await screen.findByPlaceholderText('Email Adress');
    expect(email).toBeVisible();
    userEvent.type(email, 'anemail@mail.com');
    expect(email).toHaveValue('anemail@mail.com');
    expect(email).toHaveProperty('required', true);
    expect(email).not.toHaveProperty('required', false);
    expect(email).not.toHaveValue('');
  });

  it('Renders Password input', async () => {
    const password = await screen.findByPlaceholderText('password');
    expect(password).toBeVisible();
    userEvent.type(password, '123456');
    expect(password).toHaveValue('123456');
    expect(password).toHaveProperty('required', true);
    expect(password).not.toHaveProperty('required', false);
    expect(password).not.toHaveValue('');
  });

  it('Renders Password Confirmation input', async () => {
    const passwordConfirmation = await screen.findByPlaceholderText('Password Confirmation');
    expect(passwordConfirmation).toBeVisible();
    userEvent.type(passwordConfirmation, '123456');
    expect(passwordConfirmation).toHaveValue('123456');
    expect(passwordConfirmation).toHaveProperty('required', true);
    expect(passwordConfirmation).not.toHaveProperty('required', false);
    expect(passwordConfirmation).not.toHaveValue('');
  });

  it('Renders File Form input', async () => {
    const fileInput = await screen.findByTestId('sign-up-container');
    expect(fileInput).toBeVisible();
  });

  it('Cannot render and cannot render data wothout input typed', async () => {
    axiosMock.post.mockResolvedValueOnce({ data: {} });
    const username = await screen.findByPlaceholderText('Username');
    const passwordConfirmation = await screen.findByPlaceholderText('Password Confirmation');
    const password = await screen.findByPlaceholderText('password');
    const email = await screen.findByPlaceholderText('Email Adress');
    const button = await screen.findByTestId('sign-up-btn');
    expect(username).toBeVisible();
    expect(password).toBeVisible();
    expect(passwordConfirmation).toBeVisible();
    expect(email).toBeVisible();
    expect(button).toBeVisible();
    await act(async () => {
      fireEvent.click(button);
    });
    const message = await screen.findByTestId('sign-up-errs');
    expect(message).toBeVisible();
  });

  it('Makes a call to fetching action when filling the fields', async () => {
    axiosMock.post.mockResolvedValueOnce({ data: { logged_in: false } });
    const username = await screen.findByPlaceholderText('Username');
    const passwordConfirmation = await screen.findByPlaceholderText('Password Confirmation');
    const password = await screen.findByPlaceholderText('password');
    const email = await screen.findByPlaceholderText('Email Adress');
    const button = await screen.findByTestId('sign-up-btn');
    expect(username).toBeVisible();
    expect(password).toBeVisible();
    expect(passwordConfirmation).toBeVisible();
    expect(email).toBeVisible();
    expect(button).toBeVisible();
    userEvent.type(username, 'Mario');
    userEvent.type(email, 'anemail@mail.com');
    userEvent.type(password, '123456');
    userEvent.type(passwordConfirmation, '123456');
    await act(async () => {
      fireEvent.click(button);
    });
    expect(axiosMock.post).toHaveBeenCalledTimes(1);
    expect(axiosMock.post).not.toHaveBeenCalledTimes(0);
  });
});
