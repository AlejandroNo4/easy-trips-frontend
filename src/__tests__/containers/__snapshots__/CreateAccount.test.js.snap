// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateAccount Component matches the snapshot 1`] = `
<h1
  class="session-title"
  data-testid="Sign-up-title"
>
  Sign up
</h1>
`;
